java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,427 @@
+    isTaintStepFileCreateTempFile(node1, node2)
+  }
+
+  override predicate isSink(DataFlow::Node sink) { isDeleteFileExpr(sink.asExpr()) }
+}
+
+private class TempDirHijackingFromDeleteConfig extends DataFlow3::Configuration {
+  TempDirHijackingFromDeleteConfig() { this = "TempDirHijackingFromDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) { isDeleteFileExpr(source.asExpr()) }
+
+  override predicate isSink(DataFlow::Node sink) {
+    isNonThrowingDirectoryCreationExpression(sink.asExpr(), _)
+  }
+}
+
+private class TempDirHijackingFromDirectoryCreateToUnsafeUseConfig extends DataFlow4::Configuration {
+  TempDirHijackingFromDirectoryCreateToUnsafeUseConfig() {
+    this = "TempDirHijackingFromDirectoryCreateToUnsafeUseConfig"
+  }
+
+  override predicate isSource(DataFlow::Node source) {
+    isNonThrowingDirectoryCreationExpression(source.asExpr(), _)
+  }
+
+  override predicate isSink(DataFlow::Node sink) { not safeUse(sink.asExpr()) }
+
+  override predicate isBarrierGuard(DataFlow::BarrierGuard guard) {
+    guard instanceof DirectoryCreationBarrierGuard
+  }
+}
+
+private class DirectoryCreationBarrierGuard extends DataFlow::BarrierGuard {
+  Expr fileInstanceExpr;
+
+  DirectoryCreationBarrierGuard() {
+    isNonThrowingDirectoryCreationExpression(fileInstanceExpr, this)
+  }
+
+  Expr getFileInstanceExpr() { result = fileInstanceExpr }
+
+  override predicate checks(Expr e, boolean branch) { this.controls(e, branch) }
+}
+
+/**
+ * Holds when there is an expression `unsafeUse` which is an unsafe use of the file that
+ * is not guarded by a check of the return value of `File::mkdir(s)`.
+ */
+predicate isUnsafeUseUnconstrainedByIfCheck(DataFlow::Node sink, Expr unsafeUse) {
+  exists(DirectoryCreationBarrierGuard g, MethodAccess ma |
+    any(TempDirHijackingFromDeleteConfig c).isSink(sink) and // Sink is a call to `mkdir` or `mkdirs`
+    sink.asExpr() = ma.getQualifier() and // The method access is on the same object as the sink
+    g = ma and // The guard is the method access
+    ( // TODO: Consider replacing this entire bit of logic with `TempDirHijackingFullPathInlcudingUnsafeUse`.
+      any(TempDirHijackingFromDirectoryCreateToUnsafeUseConfig c)
+          .hasFlow(sink, DataFlow::exprNode(unsafeUse)) // There is some flow from the sink to an unsafe use of the File
+      or
+      DataFlow::localExprFlow(g.getFileInstanceExpr(), unsafeUse) // There is some local flow from the passed file instance to an unsafe use
+    ) and
+    unsafeUse != sink.asExpr() and // The unsafe use is not the sink itself
+    not g.getFileInstanceExpr() = unsafeUse and // The unsafe use is not the file instance
+    not safeUse(unsafeUse) and // The unsafe use is not a safe use
+    not g.controls(unsafeUse.getBasicBlock(), true) and
+    not booleanVarAccessGuardsGuard(g) // The guard is not guarded by a boolean variable access guard
+  )
+}
```

Comment:
This is getting more and more inscrutable to reason about IMHO, not sure what to do here.

Reply:
Nothing immediately springs to mind